### PR TITLE
Readme tutorial

### PR DESCRIPTION
**Pull request by paraynaud**
First draft of the README/index and tutorial

### CODE REVIEW

**Comment by dpo** [comment]
Ok, let me know when I can make a pass over the tutorial.

**Comment by paraynaud** [comment]
I should have done a first draft today.
I am implementing new functions for the tutorial.

**Comment by dpo** [comment]
The installation instructions talk about ExpressionTreeForge and the tutorial link takes me to the ETF tutorial.

**Comment by paraynaud** [comment]
Sorry, [PartiallySeparableNLPModels.jl tutorial](https://paraynaud.github.io/PartiallySeparableNLPModels.jl/dev/tutorial/)

File: docs/src/tutorial.md [hunk 1]
```diff
@@ -1 +1,141 @@
...
+## An `NLPModel` exploiting partial separability
+PartiallySeparableNLPModels.jl defines a subtype of `AbstractNLPModel` to exploit automatically the partially-separable structure of $f:\R^n \to \R$
+```math
+ f(x) = \sum_{i=1}^N f_i (U_i x) , \; f_i : \R^{n_i} \to \R, \; U_i \in \R^{n_i \times n},\; n_i \ll n,
+```
+as the sum of element function $f_i$.
+
+PartiallySeparableNLPModels.jl relies on [ExpressionTreeForge.jl](https://github.com/JuliaSmoothOptimizers/ExpressionTreeForge.jl) to detect the partially-separable structure and defines the suitable partitioned structures, required by the partitioned derivatives, using [PartitionedStructures.jl](https://github.com/paraynaud/PartitionedStructures.jl).
+
+As a user, you only define your `ADNLPModel`:
```
Comment (dpo) [inline docs/src/tutorial.md:12]:
```suggestion
As a user, you need only define your `ADNLPModel`:
```

File: docs/src/tutorial.md [hunk 2]
```diff
@@ -1 +1,141 @@
-# PartiallySeparableNLPModels.jl Tutorial
\ No newline at end of file
+# PartiallySeparableNLPModels.jl Tutorial
+
+## An `NLPModel` exploiting partial separability
+PartiallySeparableNLPModels.jl defines a subtype of `AbstractNLPModel` to exploit automatically the partially-separable structure of $f:\R^n \to \R$
+```math
+ f(x) = \sum_{i=1}^N f_i (U_i x) , \; f_i : \R^{n_i} \to \R, \; U_i \in \R^{n_i \times n},\; n_i \ll n,
+```
+as the sum of element function $f_i$.
```
Comment (dpo) [inline docs/src/tutorial.md:8]:
```suggestion
as the sum of element functions $f_i$.
```

File: docs/src/tutorial.md [hunk 3]
```diff
@@ -1 +1,141 @@
...
+
+n = 4 # size of the problem
+model = example_model(n)
+```
+and call `PartiallySeparableNLPModel` to define a partitioned `NLPModel`:
+```@example PSNLP
+pqn_adnlp = PartiallySeparableNLPModel(model)
+```
+
+Then, you can apply the usual methods `obj` and `grad`, exploiting the partial separability, from [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl):
```
Comment (dpo) [inline docs/src/tutorial.md:32]:
```suggestion
Then, you can apply the usual methods `obj` and `grad` from [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl):
```

File: docs/src/tutorial.md [hunk 4]
```diff
@@ -1 +1,141 @@
...
+```@example PSNLP
+using NLPModels
+x = ones(n)
+fx = NLPModels.obj(pqn_adnlp, x) # compute the obective function
+```
+
+```@example PSNLP
+gx = NLPModels.grad(pqn_adnlp, x) # compute the gradient
+```
+Both `fx` and `gx` compute and accumulate respectively the element functions $\hat{f}_i$ and the element gradients $\nabla \hat{f}_i$.
```
Comment (dpo) [inline docs/src/tutorial.md:42]:
```suggestion
`fx` and `gx` compute and accumulate the element functions $\hat{f}_i$ and the element gradients $\nabla \hat{f}_i$, respectively.
```

File: docs/src/tutorial.md [hunk 5]
```diff
@@ -1 +1,141 @@
...
+In addition, a `PartiallySeparableNLPModel` store the value of each element gradient to perform partitioned quasi-Newton afterward.
```
Comment (dpo) [inline docs/src/tutorial.md:43]:
```suggestion
In addition, a `PartiallySeparableNLPModel` stores the value of each element gradient to perform partitioned quasi-Newton updates afterward. 
```

File: docs/src/tutorial.md [hunk 6]
```diff
@@ -1 +1,141 @@
...
+```
+
+```@example PSNLP
+gx = NLPModels.grad(pqn_adnlp, x) # compute the gradient
+```
+Both `fx` and `gx` compute and accumulate respectively the element functions $\hat{f}_i$ and the element gradients $\nabla \hat{f}_i$.
+In addition, a `PartiallySeparableNLPModel` store the value of each element gradient to perform partitioned quasi-Newton afterward. 
+
+```@example PSNLP
+gx == NLPModels.grad(model, x)
```
Comment (dpo) [inline docs/src/tutorial.md:46]: move this before the previous paragraph.

File: docs/src/tutorial.md [hunk 7]
```diff
@@ -1 +1,141 @@
...
+
+jumpnlp_example = jump_example(n)
+pqn_jumpnlp = PartiallySeparableNLPModel(jumpnlp_example)
+
+fx = NLPModels.obj(pqn_jumpnlp, x) # compute the obective function
+gx = NLPModels.grad(pqn_jumpnlp, x) # compute the gradient
+```
+
+## A partitioned quasi-Newton `NLPModel`
+By defining a `PartiallySeparableNLPModel` you allocate a partitioned quasi-Newton update, which is implemented in `PartitionedStructures.jl`.
```
Comment (dpo) [inline docs/src/tutorial.md:74]:
```suggestion
When defining a `PartiallySeparableNLPModel`, you allocate storage for partitioned quasi-Newton updates, which are implemented in `PartitionedStructures.jl`.
```

File: docs/src/tutorial.md [hunk 8]
```diff
@@ -1 +1,141 @@
...
+
+The Hessian approximation of each element function $\hat{f}_i (y) = (y_1 + y_2)^2$ is initially set to an identity matrix.
```
Comment (dpo) [inline docs/src/tutorial.md:76]: $\hat{f}_i$ not defined

File: docs/src/tutorial.md [hunk 9]
```diff
@@ -1 +1,141 @@
...
+  & & \left ( \begin{array}{cc}
+    1 & \\
+    & 1 \\ 
+  \end{array} \right )\\
+\end{array}
+\right ]
+```
+The accumulated matrix can be visualized with:
+```@example PSNLP
+Matrix(hess_approx(pqn_jumpnlp))
```
Comment (dpo) [inline docs/src/tutorial.md:114]: What kind of object does `hess_approx` return?

File: docs/src/tutorial.md [hunk 10]
```diff
@@ -1 +1,141 @@
...
+PartiallySeparableNLPModel(jumpnlp_example; name=:plse) # by default
+```
+The possible variants are: `:pbfgs, :psr1, :pse, :plbfgs, :plsr1` and `:plse`, see the [PartitionedStructures.jl tutorial](https://juliasmoothoptimizers.github.io/PartitionedStructures.jl/dev/tutorial/) for more details.
+
+Then, you can update the partitioned quasi-Newton approximation with the pair `x,s`:
+```@example PSNLP
+s = rand(n)
+update_nlp(pqn_adnlp, x, s)
+```
+and you can perform the partitioned matrix-vector product with:
```
Comment (dpo) [inline docs/src/tutorial.md:128]:
```suggestion
and you can perform a partitioned-matrix-vector product with:
```

File: docs/src/tutorial.md [hunk 11]
```diff
@@ -1 +1,141 @@
...
+v = ones(n)
+Bv = hprod(pqn_adnlp, x, v)
+```
+
+An in-place variant helps define a `LinearOperator` (see [LinearOperators](https://github.com/JuliaSmoothOptimizers/LinearOperators.jl)) from a `PartiallySeparableNLPModel`:
+```@example PSNLP
+using LinearOperators
+T = eltype(x)
+B = LinearOperator(T, n, n, true, true, ((Hv, v) -> hprod!(pqn_adnlp, x, v, Hv)))
+B*v
```
Comment (dpo) [inline docs/src/tutorial.md:139]: This should not be necessary. Normally, `hess_op` should "just work" if your `hprod!()` follows the NLPModels API.

**Review by dpo — CHANGES_REQUESTED** [review]

File: docs/src/tutorial.md (hunk 8 above)
Comment (dpo) [reply docs/src/tutorial.md:76]: Not really. There is $f_i$, but not $\hat{f}_i$.

File: docs/src/tutorial.md [hunk 12]
```diff
@@ -133,9 +133,8 @@ Bv = hprod(pqn_adnlp, x, v)
 
 An in-place variant helps define a `LinearOperator` (see [LinearOperators](https://github.com/JuliaSmoothOptimizers/LinearOperators.jl)) from a `PartiallySeparableNLPModel`:
 ```@example PSNLP
-using LinearOperators
 T = eltype(x)
```
Comment (dpo) [inline docs/src/tutorial.md:136]:
```suggestion
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: docs/src/tutorial.md [hunk 13]
```diff
@@ -1 +1,140 @@
...
+
+```@example PSNLP
+gx = NLPModels.grad(pqn_adnlp, x) # compute the gradient
+```
+
+```@example PSNLP
+gx == NLPModels.grad(model, x)
+```
+`fx` and `gx` compute and accumulate the element functions $f_i$ and the element gradients $\nabla f_i$, respectively.
+In addition, a `PartiallySeparableNLPModel` stores the value of each element gradient to perform partitioned quasi-Newton updates afterward.
```
Comment (dpo) [inline docs/src/tutorial.md:47]:
```suggestion
In addition, a `PartiallySeparableNLPModel` stores the value of each element gradient to perform partitioned quasi-Newton updates afterward (see below).
```

File: docs/src/tutorial.md [hunk 14]
```diff
@@ -1 +1,140 @@
...
+\left [
+\begin{array}{ccc}
+  0 & & \\
+  & 0 & \\
+  & & \left ( \begin{array}{cc}
+    1 & \\
+    & 1 \\ 
+  \end{array} \right )\\
+\end{array}
+\right ]
```
Comment (dpo) [inline docs/src/tutorial.md:110]:
```suggestion
\right ].
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: docs/src/tutorial.md [hunk 15]
```diff
@@ -1 +1,139 @@
...
+## An `NLPModel` exploiting partial separability
+PartiallySeparableNLPModels.jl defines a subtype of `AbstractNLPModel` to exploit automatically the partially-separable structure of $f:\R^n \to \R$
+```math
+ f(x) = \sum_{i=1}^N f_i (U_i x) , \; f_i : \R^{n_i} \to \R, \; U_i \in \R^{n_i \times n},\; n_i \ll n,
+```
+as the sum of element functions $f_i$.
+
+PartiallySeparableNLPModels.jl relies on [ExpressionTreeForge.jl](https://github.com/JuliaSmoothOptimizers/ExpressionTreeForge.jl) to detect the partially-separable structure and defines the suitable partitioned structures, required by the partitioned derivatives, using [PartitionedStructures.jl](https://github.com/paraynaud/PartitionedStructures.jl).
+
+As a user, you need only define your `ADNLPModel`:
```
Comment (dpo) [inline docs/src/tutorial.md:12]:
```suggestion
As a user, you need only define an `NLPModel` with an objective function implemented in pure Julia.
For instance, one may use an `ADNLPModel`:
```

**Review by dpo — CHANGES_REQUESTED** [review]

**Review by dpo — APPROVED** [review]
let's go!

**Pull request merged**